### PR TITLE
Dual Evaluation with FInAT

### PR DESCRIPTION
**Pull request by mkan88**
Performing dual evaluation in FInAT has been a long awaited feature in firedrake. This allows interaction with dual_evaluation methods in FInAT. To be used with corresponding branches in [FInAT #57](https://github.com/FInAT/FInAT/pull/57) and [TSFC #218](https://github.com/firedrakeproject/tsfc/pull/218).

### CODE REVIEW

**Comment by pbrubeck** [comment]
Could you include timing plots for the mult() method in Standalone interpolation matrix?
[fast_interp.py.txt](https://github.com/firedrakeproject/firedrake/files/4798563/fast_interp.py.txt)

**Comment by wence-** [comment]
@pbrubeck I don't think this will yet do the fast thing that you want (it just ports existing functionality right now).

**Comment by mkan88** [comment]
Currently only code restructuring have been implemented, so there shouldn't be any improvement in runtimes (yet)

**Comment by ReubenHill** [comment]
@mkan88 you can avoid unnecessary merge commits like 7f99dc3 by doing `git pull --ff`

File: firedrake/interpolation.py [hunk 1]
```diff
@@ -228,18 +238,18 @@ def _interpolator(V, tensor, expr, subset, arguments, access):
...
-        ast, oriented, needs_cell_sizes, coefficients, _ = compile_expression_dual_evaluation(expr, to_element, coords,
-                                                                                              domain=V.mesh(), coffee=False)
-        kernel = op2.Kernel(ast, ast.name, requires_zeroed_output_arguments=True)
+        ast, oriented, needs_cell_sizes, coefficients, _ = compile_expression_dual_evaluation(expr, element, coords, coffee=False)
+        kernel = op2.Kernel(ast, ast.name)
     elif hasattr(expr, "eval"):
         to_pts = []
-        for dual in to_element.fiat_equivalent.dual_basis():
+        # For FIAT elements
+        for dual in element.dual_basis():
```
Comment (ReubenHill) [inline firedrake/interpolation.py:246]: I think this line, with `element` renamed back to to `to_element` is the only change needed in `firedrake/interpolation.py` thanks to the changes introduced in #1818

File: firedrake/preconditioners/pmg.py [hunk 2]
```diff
@@ -250,14 +250,14 @@ def prolongation_transfer_kernel_aij(Pk, P1):
...
+    from tsfc.finatinterface import create_element
     from firedrake import TestFunction
 
     expr = TestFunction(P1)
     coords = Pk.ufl_domain().coordinates
-    to_element = create_base_element(Pk.ufl_element())
+    element = create_element(Pk.ufl_element())
 
-    ast, oriented, needs_cell_sizes, coefficients, _ = compile_expression_dual_evaluation(expr, to_element, coords, coffee=False)
+    ast, oriented, needs_cell_sizes, coefficients, _ = compile_expression_dual_evaluation(expr, element, coords, coffee=False)
```
Comment (ReubenHill) [inline firedrake/preconditioners/pmg.py:260]: I believe all the changes here are superseded by #1818 and should be removed from this PR.

**Review by ReubenHill — CHANGES_REQUESTED** [review]

File: firedrake/interpolation.py [hunk 3]
```diff
@@ -228,18 +238,18 @@ def _interpolator(V, tensor, expr, subset, arguments, access):
     if not isinstance(expr, firedrake.Expression):
         if expr.ufl_domain() and expr.ufl_domain() != V.mesh():
             raise NotImplementedError("Interpolation onto another mesh not supported.")
-        ast, oriented, needs_cell_sizes, coefficients, _ = compile_expression_dual_evaluation(expr, to_element, coords,
-                                                                                              domain=V.mesh(), coffee=False)
-        kernel = op2.Kernel(ast, ast.name, requires_zeroed_output_arguments=True)
+        ast, oriented, needs_cell_sizes, coefficients, _ = compile_expression_dual_evaluation(expr, element, coords, coffee=False)
```
Comment (wence-) [inline firedrake/interpolation.py:241]: The new codepath is quite a bit slower than the old one. This suggests that we need to bite the bullet and implement caching of these kernels just like we presently do for assembly kernels.

Comment (ReubenHill) [reply firedrake/interpolation.py:241]: Does the ability to store an `Interpolator` object not cut the mustard then?

Comment (wence-) [reply firedrake/interpolation.py:241]: Not really.

Comment (wence-) [reply firedrake/interpolation.py:241]: The general mantra is that the first time you run your code things might take a while, but new runs should be fast. This is now not the case for anything that does interpolation.

File: firedrake/interpolation.py (hunk 1 above)
Comment (miklos1) [reply firedrake/interpolation.py:246]: 👍 for throwing away unnecessary changes and/or superfluous if-then-else jungle.

**Comment by ReubenHill** [comment]
So somehow it seems N2div on tetrahedron cells is broken (try running `tests/regression/test_interpolation_nodes.py`). Any ideas why that might be?

**Comment by wence-** [comment]
> So somehow it seems N2div on tetrahedron cells is broken (try running tests/regression/test_interpolation_nodes.py). Any ideas why that might be?

Too little information to guess.

**Comment by ReubenHill** [comment]
> > So somehow it seems N2div on tetrahedron cells is broken (try running tests/regression/test_interpolation_nodes.py). Any ideas why that might be?
> 
> Too little information to guess.

If I run the following
```python
from firedrake import *
m = UnitCubeMesh(1, 1, 1)
V = FunctionSpace(m, "N2div", 2)
x = SpatialCoordinate(m)
expr = 2 * SpatialCoordinate(m)
v = Function(V).interpolate(expr)
```
then the generated `kernel` after `kernel = op2.Kernel(ast, ast.name, requires_zeroed_output_arguments=True)` has `len(kernel.code.domains) == 4682` and  `len(kernel.code.instructions) == 6056`. 

A similar issue occurs when interpolating into BDM (see slack discussion)
```python
from firedrake import *
m = UnitSquareMesh(1,1)
V = VectorFunctionSpace(m, "BDM", 2)
x = SpatialCoordinate(m)
expr = outer(x, x)
v = Function(V).interpolate(expr)
```
So the kernels are massive which causes kernel scheduling to take ages (credit to you for explanation - recorded here for posterity). 

Now just have to work out how to fix it. I presume this is something to do with the way `dual_evaluation` has been implemented in https://github.com/FInAT/FInAT/pull/57 at `finat/finiteelementbase.py`.

**Comment by wence-** [comment]
> Now just have to work out how to fix it. I presume this is something to do with the way dual_evaluation has been implemented in FInAT/FInAT#57 at finat/finiteelementbase.py.

I have described in that PR a bunch of things that look suboptimal to me (in terms of using ListTensor when ComponentTensor would do, and unrolling all components from VectorFunctionSpace rather than using Delta). I think those things need to be fixed.

**Comment by ReubenHill** [comment]
From Kaushik on Slack (for posterity):

> You could try re-installing islpy, as the new version is shipped with a better performing bignum package. See https://github.com/inducer/islpy/issues/28 for context. Other than that what might help is making check_bounds an optional loopy check and turn it off in firedrake when PYOP2_DEBUG is false .

**Comment by ReubenHill** [comment]
Superseded by #2115

**Pull request closed without merging**